azext_iot/custom.py

```diff
@@ -800,3 +815,279 @@ def iot_device_upload_file(client, device_id, hub_name, file_path, content_type,
+                                     attestation_type,
+                                     dps_name, 
+                                     resource_group_name,
+                                     endorsement_key = None,
+                                     certificate_path = None,
+                                     device_id = None,
+                                     iot_hub_host_name = None,
+                                     initial_twin_tags = None,
+                                     initial_twin_properties = None,
+                                     provisioning_status = None):
+    target = get_iot_dps_connection_string(client, dps_name, resource_group_name)
+    try:
+        m_sdk, errors = _bind_sdk(target, SdkType.dps_sdk)
+        
+        if attestation_type == AttestationType.tpm.value: 
+            if not endorsement_key:
+                raise CLIError('Endorsement key is requried')
+            tpm = TpmAttestation(endorsement_key)
+            attestation = AttestationMechanism(AttestationType.tpm.value, tpm)
+        if attestation_type == AttestationType.x509.value:
+            attestation = _get_attestation_with_x509_client_cert(certificate_path)
+         
+        initial_twin = _get_initial_twin(initial_twin_tags, initial_twin_properties)
+        enrollment = IndividualEnrollment(enrollment_id, 
+                                          attestation, 
+                                          device_id, 
+                                          None, 
+                                          iot_hub_host_name, 
+                                          initial_twin,
+                                          None,
+                                          provisioning_status)
+        
+        return m_sdk.device_enrollment.create_or_update(enrollment_id, enrollment, API_VERSION)
+    except errors.ErrorDetailsException as e:
+        raise CLIError(e)
+
+def iot_dps_device_enrollment_update(client, 
+                                     enrollment_id, 
+                                     dps_name,                                    
+                                     resource_group_name,
+                                     etag,
+                                     endorsement_key = None,
+                                     certificate_path = None,
+                                     device_id = None,
+                                     iot_hub_host_name = None,
+                                     initial_twin_tags = None,
+                                     initial_twin_properties = None,
+                                     provisioning_status = None):
+    target = get_iot_dps_connection_string(client, dps_name, resource_group_name)
+    try:
+        m_sdk, errors = _bind_sdk(target, SdkType.dps_sdk)
+           
+        enrollment_record = m_sdk.device_enrollment.get(enrollment_id, API_VERSION)
+        if not etag == enrollment_record['etag'].replace('"', ''):
```

Comment:
I sepecifically choose not to used the generic update, because the way this API is designed. Thare are quite a lot readonly properties in the GET response, and if I am the user, I would quite confused why I am not able to update a lot of the properties in them. That's why I have the update implemented very similar to create, to make it easier for the users.